Proposal for policy store specification

ossdhaval opened the pull request:
[Rendered](https://github.com/ossdhaval/cedar-rfcs/blob/main/text/0101-policy-store-specification.md)

ossdhaval committed:
feat: create specification for policy store

ossdhaval committed:
Update and rename 0000-policy-store-specification.md to 0101-policy-store-specification.md

emina commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,428 @@
+
+### Lack of a recommended method to co-locate and bind policies, schema, and other information
+
+Cedar does not prescribe any method or system that co-locates and binds together the information about the schema, and the policies that use that schema. In Cedar, a policy's binding with its schema happens at the time of evaluation. This approach gives flexibility to combine any schema with any policy at evaluation time. But in practice, more often than not, the policies are targeted to a specif ...
```
Note that policy evaluation is independent of schemas in Cedar.  That is, you don't need any schema at all to evaluate policies.  You only need schemas to validate or analyze policies---or, if you are (optionally) using schema-based entity parsing, to parse entities.

So, I think we need to rephrase the motivation here to clarify this:  there are no required dependencies on schemas at authorization (run) time.

nynymike replied:
Yes, we are doing "schema-based entity parsing" in the Cedarling based on the rules for a trusted issuer, as mentioned in the docs [here](https://docs.jans.io/stable/cedarling/cedarling-policy-store/#claim-mapping).

cdisselkoen replied:
You can pass them in `entities` if you do not pass a schema.  (And if you don't have action groups, I believe you don't need to pass them in either `entities` or schema)

emina commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,428 @@
+
+## Alternatives
+
+Not applicable
```
The alternative is to maintain status quo and not provide a policy store specification as part of Cedar.

patjakdev replied:
My feedback is that I don't really understand why the community needs an authoritative specification for the "policy store" concept. The specification you've created seems fine for your specific use case, but I don't really see a need for the Cedar team to dictate (and maintain into the future) such a specification.

Are you looking to interoperate with policy stores generated by third parties or something?

Swolebrain replied:
There's one alternative I've outlined above. In general, alternatives are always applicable. For instance with the alternative I outlined, we make it more difficult to cryptographically sign the state of a particular policy store, but we can support more use cases. Another alternative is to specify the cedar schema in a way that's driven by the API spec. It can be versioned and signed in that way. We have a reference implementation for how an OpenAPI spec can be made to express a cedar schema here: https://github.com/cedar-policy/cedar-authorization

Regarding the mapping of issuers, I believe what Phil meant is that we have oidc hardcoded into this spec without a way to specify other types of identity sources, like SAML or any future ones.

windley replied:
Late to the party here... 

I don't know that we need a specification that binds Cedar to a specific format, but I do see value in having well-thought-out formats for how to package and move schema and policies (which is what I see this being about primarily).

RoyalCookie commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,428 @@
+    - `schema`: base64-encoded Cedar schema used by policies in this store. There can be only one schema per policy store.
+    - `default_entities`: A mapping of entity IDs to base64-encoded JSON entity data. 
+
+##### Trusted issuers
```
I think attempting to defined all these new concepts in the initial RFC might be a mistake, I'd expect some pushback to simply introducing the Policy Store concept to Cedar.

But I think every AVP/OIDC/OAuth specific attribute outlined in the RFC  requires it's own discussion so it might be more productive to simplify the scope of this RFC.

Limit it specifically to introducing the `Policy Store` concept where only the core attributes are defined such as `description`, `policy_content`, `creation_date?`, but allows for additional user defined  attributes perhaps?

Swolebrain replied:
I completely agree we should leave out identity-related portions of the specification for the time being.

There's a tremendous amount of complexity in coming up with a general-enough abstraction to support a majority of oidc providers, due to the fact that they all deviate from the oidc spec in slightly different ways. For example:

* The concept of "groups" is modeled differently. Okta models it as a "Groups" claim that can only be present in identity tokens. Ping models it as a "groups" claim that is present in both identity and access tokens. Cognito models it as a "cognito:groups" claim that can be present in either token. Auth0 models something similar as a "permissions" claim. 
* Different levels of support for JWE token encryption
* Different treatments for `client_id` claim. The one example I can remember is that Auth0 sometimes uses an `azp` claim (depending on which type of "auth0 application" the token comes from) instead of `client_id` to represent the client id.
* Can't find the details right now but if my memory doesn't fail me, there's also different formats for some standard claim values (where one idp will want a space-separated string another one might want a string array).

Beyond that, this proposal doesn't leave room to support other identity sources, like for example SAML, or any future specs that might come out (perhaps for MCP's or what not).

nynymike replied:
> And while it is supported by Amazon Verified Permissions, it is limited to one Trusted Issuer at the moment.
> 

So if we could define the spec in a way that supports AVP's one Trusted Issuer, but open the possibility that we'll need to make policies about more then one issuer... that would be a win right?  

Is it really so future looking to say that we may need to write policies about tokens from more then one issuer? I would say we're already seeing these requirements in the AI world.

Swolebrain commented on text/0101-policy-store-specification.md:
````diff
@@ -0,0 +1,428 @@
+```
+{
+    "cedar_version": "4.4.0",
+    "policy_stores": {
+        "9e96b204911d1c3": {
+            "name": "Acme Analytics Web Application",
+            "description": "Acme's premier tool for counting narwhals, doves and gorns.",
+            "policies": {"a270f688133cd00": "base64-content",...},
+            "trusted_issuers": {...},
+            "schema": "base64-content"
+            "default_entities": {"1694c954f8d9": "base64-content",...}
+        }
+    }
````
While I think the idea of standardizing on a policy store format is fantastic, and I thank you for your contribution, I have several objections to this format. I will try to organize them by theme below.

I will use the term **builder** to refer to people like us, who build applications using Cedar, and use the term **end user** to refer to the users using those applications.

### 1. Use case support
There are at least two major use cases that are left out, not possible to support by the format outlined in this proposal: multi-tenancy and dynamic policy stores (as well as use cases which have both attributes). This format is good for IaC-driven policy stores but it is not good for dynamic policy stores. 

To summarize, I will use as an example a use case that requires both: dynamic policies and multi-tenancy. Let's say you're a builder working on ClickDown.com, an imaginary productivity app that allows you to manage tasks, create and share docs, manage content, and even live chat with coworkers. You're using Cedar to model the authz model of ClickDown. With Cedar, when building a multitenant app like ClickDown, the best practice is to have one policy store per tenant. This allows for permission isolation and avoids accidental collisions. Since your application is very successful and you market with the freemium model, your app has 80 million users. That makes this JSON file be somewhere north of 2.4 terabytes (assuming 30kb per policy store, which is actually low). 

Additionally, each tenant of ClickDown has many end users. Admin end-users of the ClickDown app have the ability to create other users. Since you user Cedar, there's a lot of flexibility in what you can do in your "Create user" flow. You provide your tenant admins with a flow such that they can select all the actions and resources the new user is allowed to act on. One or more cedar policies are created in the backend. So when you create a new user `ClickDown::User::"123123"`, there are associated policies like `permit(principal == ClickDown::User::"123123", action...);` generated specific to that user. Under this use case (which is common), each tenant will have policy stores that will grow arbitrarily. One tenant by itself might have 100,000 users in it, with 100,000 policies. It's impractical to version the whole policy store when one single policy was created or updated.

### 2. Format serialization

If a policy store descriptor format is to exist, one of its uses should be to serialize dynamic policy stores for the purposes of disaster recovery. Serializing a multi-gigabyte file in json is needlessly difficult. Since you can't load it all into memory (because you run oom) you need to write a streaming serializer. This needlessly increases the complexity of encoding this format.

It's much easier to compare the git `HEAD` to `HEAD~1` and only deploy the files that were changed.

De-serializing the format to compare it to what's already deployed has similar streaming complications.

### 3. Support for templates

If a policy store descriptor format is to exist, it should support policy templates and template linked policies.

### My proposal

I would recommend that instead of this, the policy store format should:

1. Encompass a single policy store
2. Store units in the most atomic manner possible, so they can be versioned independently and thus deployed independently.

One structure could be like so:

```
└── root/
    ├── policyStoremetadata.json                          # Export metadata (cedar version, name, description, etc)
    ├── schema.cedarchema                 # Schema definition for the policy store
    ├── static_policies/                       # Directory for standalone policies
    │   ├── policy_1.cedar                     # Individual Cedar policy files, filename should be policy id
    │   ├── policy_2.cedar
    │   └── policy_n.cedar
    ├── templates/                             # Directory for templates and their linked policies
    │   ├── template_1/                      # Folder should be template id
    │   │   ├── template.cedar         
    │   │   ├── link_1.json                    # Batch of template variable combinations representing template linked policies
    │   │   ├── link_2.json                    
    │   │   └── link_n.json
    │   ├── template_2/
    │   │   ├── template.cedar
    │   │   └── link_1.json
    │   └── template_n/
    │       └── template.cedar
    └── manifest.json   
```

Note there's a `manifest.json` file which can be used to list all the contents of a policy store. It would end up being something reminiscent of an `info.plist` file in an iOS app or a progressive web app's `manifest.json` file. I haven't thought through this in detail, but one pro of having a manifest is that it _should_ be easier to do static analysis validation on the manifest. For example, if the manifest says that the policy store should perform validation at authorization time, then there must be a schema file present, it must be parseable, and all the policies named in the manifest must be validated against the schema.

nynymike replied:
I actually quite like this design... I even like the use of templates... for those who don't want it, don't use it.

Swolebrain commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,428 @@
+
+### Strong versioning
+
+* For forensic analysis of an audit log, you must know the exact version of the policies and schema in use at the time of the decision. 
+* Versioning provides an opportunity for review and governance. For example, the CISO may want to review the policies of a certain application and then review all subsequent changes. Versioning aligns with existing change management and CI/CD pipelines.
+* Strong versioning aligns with cloud native declartive syntax goals, making the deployment of new code and rollback to old code easier.
```
Using a single file makes these 3 versioning concerns difficult to support:

* For forensic analysis, under the context of multi-tenancy and dynamic policies, if tenant1 changes one policy, that creates a whole new version of the policyStores file. So tenant 700 sees a different policy store version in audit logs even though they've changed no policies
* If a policy store has a thousand policies and only one policy is changed, now my IAC deployer needs to download the whole policy store, diff it against the file, and change what's necessary to change. That, or put write pressure on the DB re-writing everything. Neither is preferrable over updating only what needs to be updated, which can be accomplished by versioning at a more granular level.
* See above, it's much easier to deploy changes and roll back if you know what specific things changed in each policy store.

nynymike replied:
I don't actually agree that every policy change requires a new policy store version. If you have 700 policies, you edit 50 of them, and then submit a PR for approval, and once approved... it gets released as a new policy store version.

If your app has dynamic policies, then a policy store version won't make sense for you. I think the individual policy version is just as good as long as you have a way to create a chain of custody for the audit log.

We will try to flush out these examples in the revision...

ossdhaval commented:
Thank you everyone for your time and input. We have revised the RFC to address the comments. Please follow the link in the PR description to read the updated version. 

#### Major changes

- Moving from a single JSON file structure to a developer-friendly directory format. For suggesting this idea, @Swolebrain has been added as a contributor. :-)
- Each policy, template, and entity is stored as an individual, human-readable file. No base64-encoded strings. Greatly improving version control, IDE support, and collaborative workflows
- Design allows for multiple trusted issuers
- For deployment or distribution, the proposal introduces a standardized .cjar archive format that preserves the structured directory
- The RFC details file conventions, extensible metadata, and open questions while explicitly outlining drawbacks and alternatives


And many more changes.

Looking forward to your inputs on the revised RFC.

ossdhaval committed:
fix: address review comments

Signed-off-by: ossdhaval <343411+ossdhaval@users.noreply.github.com>

benkehoe commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,595 @@
+
+The proposal in this RFC targets the following constraint:
+
+### Lack of a recommended method to co-locate and bind policies, schema, and other information
```
I'm struggling with why this makes sense to include as part of Cedar proper.

> policies are targeted to a specific schema

This might be addressable with a much more narrow proposal, like a well-known policy annotation for linking the policy to a schema version.

> In the audit logs to create a cryptographic chain of custody, it's critical to know the exact version of the policies and the schema used at evaluation time. For example, when using schema-based entity parsing.

This is a very use case-specific assumption, that the policies as collected and schema are accessed via a consistent snapshot for authorization decisions, and that schema is used at authorization time. Cedar itself does not define either of these things.

The rest of the motivation deals with interoperability of authorization systems that are built _using_ Cedar, as opposed to the interoperability of Cedar itself. It doesn't make sense to me, for example, that Cedar would define a format that includes trusted issuers, which are not represented anywhere within Cedar itself.

My take here is that a better path might be establishing a separate effort (within or alongside the Cedar project) to define (optional) standards for authorization systems built on top of Cedar. A policy store spec is one aspect. Another could be, for example, a standardized way to provide entities in a call to an AuthZEN-compliant authorization endpoint.

nynymike commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,595 @@
+
+#### Default Entities
+
+Default entities are stored as individual JSON files in the `entities/` directory. These files capture static resource entities, such as organization information that doesn't frequently change.
```
Ok, hence the `default_` part of the claim :-)

blubben commented on text/0101-policy-store-specification.md:
```diff
@@ -0,0 +1,595 @@
+policy-store/
+├── metadata.json                    # Policy store metadata
+├── manifest.json                    # Inventory of files
+├── schema.cedarschema               # Cedar schema file
```
I'd suggest making the schema a subdirectory of its own. It can be convenient to break a schema up by namespace (and sharing schema fragments across projects.)

ossdhaval committed:
fix: revise RFC to address review comments

ossdhaval commented:
Hello everyone, 

This RFC has been updated to address the feedback comments. 

**Highlight of the changes:**

- Introduced the new `manifest.json` file that provides an inventory of all files in the policy store for validation and integrity checking
- Added samples to cover action groups
- Expanded details on `Default entities`
- Added JSON Schema for trusted issuers and entity files

Please share your views for the updated RFC.

nynymike commented:
We made some changes to address the feedback, like simplifying the `trusted_issuer` metadata a little and adding another example to the entity.  I hope we can make progress on this RFC soon.

nynymike commented:
https://aws.amazon.com/blogs/aws/amazon-bedrock-agentcore-adds-quality-evaluations-and-policy-controls-for-deploying-trusted-ai-agents/
<img width="888" height="120" alt="image" src="https://github.com/user-attachments/assets/64f5fb37-5471-486f-a4db-124efbf5e53a" />

Governance needs to be interoperable. While CNCF community contributors like Amazon are developing interesting services, their claims to audit and compliance teams are hard to accomplish without specific versioned policies. We need to know what version of policy is in the execution environment, and how it got there. And we need continuous evaluation according to an assessment procedure, without which, the policy is merely a recomendation. Lack of a versionable binary like this makes Cedar impossible to govern in an interoperable way.

nynymike commented:
Duly noted. I'm in the process of proposing a "GovOps" working group at OpenSSF, which could be a home for interoperability related standards.

nynymike commented:
I'm proposing this as an AuthZen Standard. See draft [here](https://htmlpreview.github.io/?https://github.com/nynymike/AuthZen_Policy_Store/blob/main/draft-schwartz-authzen-policy-store.html)

benkehoe commented:
I think AuthZen makes a ton of sense for this.

ossdhaval committed:
Merge branch 'cedar-policy:main' into main

ossdhaval commented:
We are closing this RFC in light of the proposed AuthZen Standard. 

Thank you everyone for chipping in. Your inputs have helped us shape the original idea. :pray:
